specs/agents/tracing-instrumentation-aws-lambda.md

```diff
@@ -0,0 +1,127 @@
+
+- API Gateway V1 
+- API Gateway V2 
+- SQS
+- SNS
+- S3
+
+If none of the above apply, the fallback should be a generic instrumentation (as described above) that can deal with any type of trigger (thus capturing only the minimal available information).
+
+### API Gateway (V1 & V2)
+There are two different API Gateway versions (V1 & V2) that slightly differ in the information (`event` object) that is passed to the Lambda handler function. 
+
+With both versions, the `event` object contains information about the http request.
+Usually API Gateway-based Lambda functions return an object that contains the HTTP response information.
+The agent should use the information in the request and response objects to fill the HTTP context (`context.request` and `context.response`) fields in the same way it is done for HTTP transactions.
+
+In addition the following fields should be set for API Gateway-based Lambda functions:
+Field | Value | Description 
+---   | ---   | ---
+`faas.trigger` | `http` | 
+`faas.execution` | e.g. `123456789` | Use the API Gateway request ID.
+
+With both API Gateway versions the `event` objects (may) contain a timestamp of the original API Gateway request.
+This information should be used to [adapt the transaction and spans structure](#init-spans) of the Lambda invocation to represent the initialization time.
+
+### SQS
+
+TODO: How to deal with batch messages
+
+In addition the following fields should be set for SQS-based Lambda functions:
+Field | Value | Description 
+---   | ---   | ---
+`faas.trigger` | `pubsub` | 
+`faas.execution` | e.g. `someMessageId` | Use the SQS message ID.
+`context.message.queue` | e.g. `arn:aws:sqs:us-east-2:123456789012:my-queue` | The SQS event source ARN.
+`context.message.age` | e.g. `3298` | Age of the message in milliseconds. `current_time` - `SentTimestamp`, if SentTimestamp is available. Can be retrieved from message attributes with key `SentTimestamp`.
+`context.message.body` |  | The message body, only captured if body capturing is enabled in the configuration.
+`context.message.headers` |  | The message headers / attributes, only if capturing headers is enabled in the configuration.
+
+### SNS
+
+### S3
+
+## Init spans
+With both API Gateway versions the `event` objects (may) contain a timestamp of the original API Gateway request. The difference between the API Gateway timestamp and the actual start timestamp of the Lambda function is a good estimate for the initialization phase of the lambda function. 
+The agents should use this information (if available) to adapt the transaction and spans structure of the Lambda invocation to represent the initialization time:
+
+```
+...
+  [XXXXXXXXXXXXX POST api.gateway XXXXXXXXXXXXXXXXXXXXX]   // client calling the API Gateway
```

Comment:
- It's sort of implicit in the waterfall diagram, but should we call out explicitly that  transaction starting should include any tracestate and traceparent headers that are passed along via the the `event` argument?

Reply:
Added a view words on `traceparent` and `traceparent` retrieval for relevant trigger types (API Gateway, SQS, SNS)